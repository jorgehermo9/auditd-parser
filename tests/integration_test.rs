use std::{fs, path::PathBuf};

use auditd_parser::{AuditdRecord, ParserError};
use erased_serde::Serialize;
use rstest::rstest;

#[rstest]
fn test_log_data(#[files("tests/data/**/*.log")] log_file: PathBuf) {
    let log_file_path = log_file.to_string_lossy().into_owned();
    let log_file_content = fs::read_to_string(log_file).unwrap();
    let logs = log_file_content
        .lines()
        .enumerate()
        .map(|(index, line)| (index + 1, line))
        .filter(|(_line_number, line)| !line.is_empty() && !line.starts_with('#'))
        .collect::<Vec<_>>();

    for (line_number, log) in logs {
        let maybe_record = log.parse::<AuditdRecord>();
        let location = format!("{log_file_path}:{line_number}");
        let log_identifier = get_log_identifier(log, &location);

        // Unwrap the Ok variant but leave the Err variant as-is
        let result: Box<dyn Serialize> = match maybe_record {
            Ok(record) => Box::new(record),
            Err(error) => Box::new(Err::<AuditdRecord, ParserError>(error)),
        };

        insta::with_settings!(
            {
                info => &log,
                description => location,
                snapshot_suffix => log_identifier,
            },
            {
                insta::assert_json_snapshot!(result);
            }
        );
    }
}

fn get_log_identifier(log: &str, location: &str) -> String {
    let log_md5 = md5::compute(format!("{log}{location}"));
    format!("{log_md5:x}")
}
