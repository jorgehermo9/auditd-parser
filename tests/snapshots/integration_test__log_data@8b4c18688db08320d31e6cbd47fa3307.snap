---
source: tests/integration_test.rs
description: "/home/jorge/github/auditd-parser/tests/data/linux-audit-parser-rs/line-crypto-session.log:1"
expression: result
info: "type=CRYPTO_SESSION msg=audit(1724778544.061:594047): pid=21702 uid=0 auid=4294967295 ses=4294967295 msg='op=start direction=from-client cipher=chacha20-poly1305@openssh.com ksize=512 mac=<implicit> pfs=ecdh-sha2-nistp521 spid=21704 suid=74 rport=46984 laddr=10.10.10.10 lport=22  exe=\"/usr/sbin/sshd\" hostname=? addr=10.11.11.11 terminal=? res=success'"
---
{
  "record_type": "CRYPTO_SESSION",
  "timestamp": 1724778544061,
  "id": 594047,
  "fields": {
    "auid": null,
    "msg": {
      "addr": "10.11.11.11",
      "cipher": "chacha20-poly1305@openssh.com",
      "direction": "from-client",
      "exe": "/usr/sbin/sshd",
      "hostname": null,
      "ksize": "512",
      "laddr": "10.10.10.10",
      "lport": "22",
      "mac": "<implicit>",
      "op": "start",
      "pfs": "ecdh-sha2-nistp521",
      "res": "success",
      "rport": "46984",
      "spid": "21704",
      "suid": 74,
      "terminal": null
    },
    "pid": "21702",
    "ses": "4294967295",
    "uid": "root"
  }
}
